### ISSUE TITLE
Improve timings of basic qgis_process commands

### ISSUE THREAD

**Issue by florisvdh** [qgis/QGIS#54563]
### Feature description

It would be great if a few simple `qgis_process` commands could be made to run (much) faster. On my laptop (CPU: 11th Gen Intel Core i5-1135G7) current timings for following commands are approximately:

1. `qgis_process`: 0.65 - 0.69 s
2. `qgis_process --help`: 0.83 - 0.87 s
3. `qgis_process --version`: 0.84 - 0.88 s
4. `qgis_process plugins --json`: 1.58 - 1.63 s

These timings were obtained by iteratively running commands like `/usr/bin/time -f %e qgis_process --version > /dev/null` in bash.

### Additional context

Commands 1, 3 and 4 are called when loading R package {[qgisprocess](https://github.com/r-spatial/qgisprocess)}, in order to validate the `qgis_process` path, version and plugin list that are stored as metadata in a persistent package cache. The three commands add up to > 3 seconds, while the R package code needs < 200 ms to load.

I'm wondering whether some or all of the above listed `qgis_process` commands could be made faster, as this could much shorten the time for the R package to load.

The R package cache essentially stores the algorithm metadata (as obtained with `qgis_process list --json`, taking about 1.80 s), next to the aforementioned `qgis_process` metadata. If we would not be using a package cache, commands 1, 3, 4 plus `qgis_process list --json` would need to be called each time.

So the tradeoff of caching vs not caching would already improve if e.g. commands 1 and 3 (and ideally 4) ran much faster.

**Comment by nyalldawson** [comment]
@florisvdh I had a quick look into this. There's certainly some room for improvements here.

From my initial findings there's a few opportunities to improve this:

- One big startup cost is building the default symbol style library. This could potentially be deferred until it's actually required (i.e. not when running the basic qgis_process "metadata" style commands)
- Similarly another big startup cost is setting up the QGIS authentication subsystem. This isn't required until we actually DO stuff, so again I think there's opportunity to defer this initialization until actually needed.
- There's some recent startup costs relating to setting migration which account for the bulk of the startup time for a raw qgis_process command. Should be fixable with some improved logic about when the migration is actually required.

When we look at a command like "qgis_process list", there's a few more culprits which spring up:

- One (likely unavoidable) cost is the initialisation of Python itself. (You can avoid this by passing `--no-python`, but that limits the available algorithms to just c++ ones)
- There's a significant startup cost associated with each Processing model the user has in their profile. There's likely room for improvement here.
- There's a bunch of time spent in Python code on Processing initialization. More work would be needed to break this down (eg utilising https://docs.python.org/3.12/howto/perf_profiling.html ). 
- One known signfiicant cost (especially on Windows) is the parsing of all the Grass description txt files. This is a HUGE component to the overall QGIS startup cost on Windows. Here I'd suggest we replace all the hundreds of .txt files with one single unified JSON description file, so that only one file open and read is required. 

In short, I believe there's substantial improvements possible!

**Comment by florisvdh** [comment]
Thank you @nyalldawson for having a look into this!

Maybe `qgis_process plugins` has similarities to `qgis_process list` in that it needs Python to initialize (I think). Only the _former_ is called when loading the R package. Of course if both would be fast, it might render package caching unnecessary, which simplifies things. In the other case (using the R package cache), most gain in R package loading time would come from speeding up `qgis_process plugins`, next to `qgis_process` and `qgis_process --version`. Only those three commands are called to validate the persistent R package cache.

> parsing of all the Grass description txt files

Yeah, that is definitely something confined to `qgis_process list`; should not happen in `qgis_process plugins`. You mentioned the problem in the GRASS GIS community meeting in June (minutes in https://github.com/OSGeo/grass/discussions/3047; see paragraph _Algorithm definition and description files_). The solution of holding them in one file appears much better indeed.

**Comment by florisvdh** [comment]
Adding more related PRs (next to the 6 above) for historical reference:

- Optimise QGIS startup by lazy-loading symbol styles: https://github.com/qgis/QGIS/pull/55793
- Avoid loading plugins wherever possible (affecting timing of above mentioned `qgis_process list --json`): https://github.com/qgis/QGIS/pull/55794
- Optimise QGIS startup by improving settings migration logic: https://github.com/qgis/QGIS/pull/55777
